lib/OpenQA/Schema/Result/JobModules.pm

```diff
@@ -235,4 +235,33 @@ sub save_results {
     path($self->job->result_dir, 'details-' . $self->name . '.json')->spurt(encode_json($results));
 }
 
+# Collect text output into JSON. save_results() is called too early for this.
+sub finalize_results {
+    my ($self) = @_;
+
+    my $dir = $self->job->result_dir();
+    return unless $dir;
+    my $file    = path($dir, "details-" . $self->name . ".json");
+    my $tmpfile = $file->sibling($file->basename . '.tmp');
+    return unless -e $file;
+    my $results;
+
+    eval {
+        $results = decode_json($file->slurp);
+        $results = {details => $results} if ref($results) eq 'ARRAY';
+
+        for my $step (@{$results->{details}}) {
+            next unless $step->{text};
+            my $txtfile = path($dir, $step->{text});
+            next unless -e $txtfile;
+            $step->{text_data} = $txtfile->slurp;
+        }
+
+        $tmpfile->spurt(encode_json($results));
+        rename $tmpfile->to_string, $file->to_string;
+    };
+
+    eval { $tmpfile->remove if -e $tmpfile; };
```

Comment:
Don't just pre-emptively use `eval`. If you're expecting a specific exception here then handle that case gracefully, but don't hide everything.

Reply:
I think this eval still deserves a handler - even if it would just log the message. And I'm nevertheless wondering whether we should mark the Minion job as failed *in the end* when any of these error handler were entered. Otherwise we'll probably never notice when there's a problem.

Reply:
The second eval is just a post-fail cleanup. It does nothing if the first `eval` succeeds. Do we really need `log_error "Cleaning up after the error that was already logged above also failed"`?

Reply:
Mh… ok in this case one of the error would likely be sufficient.

Reply:
That explanation doesn't make much sense to me. What is the actual purpose of this cleanup code? Why would the temp file still exist here? Why do you expect removing it to fail? It looks to me like if the first `eval` succeeds, then this would do nothing since the tempfile is gone. Do you expect the first `eval` to fail and leave a temp file behind that needs to be cleaned up? Why would the process be unable to remove it? Do you expect permissions to change? Why not do a `-w` check then?

I'm probably missing something here, but i can't imagine there is a good reason to have an `eval` without exception handling.

Reply:
@kraih I expect the code to work without any errors. But if it doesn't, there are 3 major problems I have to deal with:
1. Don't corrupt the original JSON file.
2. Don't leave any garbage behind.
3. Don't throw any exceptions out of the function because they're totally irrelevant. No data was lost or damaged, nobody cares.

If you don't understand how to do error handling properly, that's your problem, not mine.

Reply:
Point 1 seems irrelevant here, it's just a temp file. And point 3 you still left unanswered, what exception do you expect that makes the code fail but that's totally irrelevant. If the file system is that broken that you can't delete files you created that would seem like something i'd like to see as an error for in the job result (instead of just finding garbage files without any clue where they came from). That leaves only point 2, and that's just the reason for the `->remove` call, not a justification to wrap it in an `eval`. A `-w` check would do the job just as well.

Please refrain from ad hominem attacks, i'm actually a domain expert.

Reply:
I'm not trying to give you a hard time here, i just need to know your intention to be able to tell you how to solve the problem correctly. Just looking at the code it appears that a scope guard would be the best solution and what i would want to see here.
```perl
eval {
  ...
  my $tmpfile = tempfile(DIR => $file->dirname);
  ...
  $tmpfile->move_to($file);
};
```

Reply:
> Point 1 seems irrelevant here, it's just a temp file. And point 3 you still left unanswered, what exception do you expect that makes the code fail but that's totally irrelevant. If the file system is that broken that you can't delete files you created that would seem like something i'd like to see as an error for in the job result (instead of just finding garbage files without any clue where they came from). That leaves only point 2, and that's just the reason for the `->remove` call, not a justification to wrap it in an `eval`.

If the code can't delete files, you ***WILL*** see it in the log. It'll be the exact same error that was already logged when `$tmpfile->spurt()` or `rename()` failed. So there's no point logging it twice. Especially when you can expect hundreds of duplicate errors because I'm deliberately keeping the minion job going until it cycles through all test modules.

The reason for wrapping `$tmpfile->remove()` in an `eval` block is that I don't want to terminate the minion job if cleanup fails. The failure could be a temporary hickup.

`tempfile()` creates the file with permissions `0600`. I'd have to run `chmod()` in addition to `rename()`. And no, I will **NOT** use `move_to()` here.

Reply:
Then all of your problems would be solved by using a scope guard there. No need for ugly hacks.

Reply:
> Then all of your problems would be solved by using a scope guard there. No need for ugly hacks.

And risk that the scope guard behavior changes in the future and it starts throwing bogus errors when it fails to remove the renamed file? No, thanks. Do you understand that the only thing your "improvements" would add to the code is technical debt? The code is working as intended. Period.

Reply:
That's a silly line of argumentation, how do you know the `$file->sibling` call above you did not wrap in an `eval` won't change? I control that method upstream, i could change it right now just to mess with you. I won't of course, but all the methods we are talking about here are controlled by me upstream, and i'm just trying to tell you how to use them in the safest way possible.

Anyway, it doesn't matter because scope guards work exactly in the way you need them to.
```
$ perl -MMojo::Util=scope_guard -E 'eval { my $guard = scope_guard sub { warn "warning from scope guard"; die "exception from scope guard" }; die "real exception" }; warn "eval error: $@" if $@;'
warning from scope guard at -e line 1.
eval error: real exception at -e line 1.
```
Just because you don't understand a concept yet that doesn't make it technical debt. Scope guards are a standard Perl pattern, and i'm happy to explain it to you in more detail if needed.

Reply:
> That's a silly line of argumentation, how do you know the `$file->sibling` call above you did not wrap in an `eval` won't change? I control that method upstream, i could change it right now just to mess with you. I won't of course, but all the methods we are talking about here are controlled by me upstream, and i'm just trying to tell you how to use them in the safest way possible.

I would write this PR exactly the same way even if I myself were the upstream maintainer of the `Mojo` library. The code is as simple as it can be and any extra abstractions would only make it more complex.